<a href="https://colab.research.google.com/github/NatFT/PythonCourse/blob/main/Ex_Working_with_Strings.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# Working with Strings
---

The pandas library has a similar set of string functions to those available in python generally.  Because we often want to perform operations on a whole series of data values in a dataframe, we can use pandas string functions to do this:



### Creating a series from a pandas column
---

A column of data is treated by pandas as a Series.  There is a set of functions that you can access for working with a **series** (just one column from the data table).

To get a 'series' from a dataframe, you would split the column from the rest of the dataframe by taking a copy of it and storing it in a new variable (which is very similar to a list).

The examples below show what you can do with a Series rather than a whole table.

To get a column of data as a series:

```
series_data = df['date']
```
or
```
price_series = df['price']
```
where 'date' or 'price' are the names of the columns in the dataframe

### Splitting data 
---

Series.str.split() *to split a column's strings into components*    
Series.str.get() *to get one of the components after the split*  

You can **daisychain** these together:   

`Series.str.split().str.get()`

* `split()` will split by white space unless specified, for example if you wanted to split by "/" you would use `split("/")`  

* `get()` requires a parameter of the value position of the string you would like to 'get'. If you want the first word eg 1999, use `get(0)`.


*Hint: remember to save your result into a new column* 

### Exercise 1 strings
---

Let's use the data set 'Housing in London' at 'https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv'

The date, in this dataset is a string.   To filter for a particular year, we will need to extract the first four letters as a substring.  We can create a new column called **year**, which just contains the year, stored as a number.

The date is written in the format yyyy-mm-dd.  We can split the year around the '-' and then use the first component.


Create a function called **get_year()** which splits the data from the date column, and creates a year column with just the year before returning the year column.


**Test output**:  

```
0       1999
1       1999
2       1999
3       1999
4       1999
        ... 
1066    2019
1067    2019
1068    2019
1069    2019
1070    2019
Name: year, Length: 1071, dtype: object
```

In [ ]:
import pandas as pd

housing_df = pd.read_csv('https://raw.githubusercontent.com/futureCodersSE/working-with-data/main/Data%20sets/housing_in_london_yearly_variables.csv')

print(housing_df.head())
print(housing_df.columns)

def get_year():
  # add code below to return a new series created from the year column, with just years 
  year = housing_df['date'].str.split('-').str.get(0)
  display(year.head())
  return year

#get_year()


# run and test if returned series is of correct length and has correct first row 

actual_len = len(get_year())
actual_value = get_year().iloc[0]
expected_len = 1071
expected_val = "1999"

if actual_len == expected_len and actual_value == expected_val:
  print("Test passed expected length 1071 and first value 1999 and got", actual_len, actual_value)
else: 
  print("Test failed expected length 1071 and first value 1999 and got length", actual_len, "value", actual_value)




        code                  area        date  median_salary  \
0  E09000001        city of london  1999-12-01        33020.0   
1  E09000002  barking and dagenham  1999-12-01        21480.0   
2  E09000003                barnet  1999-12-01        19568.0   
3  E09000004                bexley  1999-12-01        18621.0   
4  E09000005                 brent  1999-12-01        18532.0   

   life_satisfaction mean_salary recycling_pct  population_size  \
0                NaN       48922             0           6581.0   
1                NaN       23620             3         162444.0   
2                NaN       23128             8         313469.0   
3                NaN       21386            18         217458.0   
4                NaN       20911             6         260317.0   

   number_of_jobs  area_size  no_of_houses  borough_flag  
0             NaN        NaN           NaN             1  
1             NaN        NaN           NaN             1  
2             NaN        NaN 

0    1999
1    1999
2    1999
3    1999
4    1999
Name: date, dtype: object

0    1999
1    1999
2    1999
3    1999
4    1999
Name: date, dtype: object

Test passed expected length 1071 and first value 1999 and got 1071 1999


### Exercise 2
---

In exercise 1 you have extracted the year, but it's dtype is 'object' (it is still a string).  You can convert to integer by adding  .astype(int) to the daisychain.

Create a new function called **get_int_year()**, `return` the year column with values of type int. 

**Test output**:  

```
...
Name: year, Length: 1071, dtype: int64
```



In [ ]:
def get_int_year():
  # add code below to return a year column where values are of integer type
  year = housing_df['date'].str.split('-').str.get(0).astype(int)
  display(year.head())
  return year





# run and test if your returned series is of type int

import numpy as np

actual = get_int_year().dtype
expected = np.int64

if actual == expected:
  print("Test passed", actual)
else:
  print("Test failed, expected", expected, "got", actual)

0    1999
1    1999
2    1999
3    1999
4    1999
Name: date, dtype: int64

Test passed int64


### Exercise 3
---

All the areas in the data set are in lower case.  To prepare the data for reporting, you may want to capitalise.  Use .str.title() to do this.

Create the function **get_title_areas()** to do this. `Return` the newly capitalised area column 

In [ ]:
def get_title_areas():
  # add code below to capitalise the first letter of each string in the column 'area'
  area = housing_df['area'].str.title()
  return area






# run and test if the first row of the area column is now correct 

actual = get_title_areas().iloc[0]
expected = "City Of London"

if actual == expected:
  print("Test passed", actual)
else:
  print("Test failed, expected", expected, "got", actual)

Test passed City Of London


### Exercise 4 - Filter all areas to find all with 'And' in the name
---

Create a function called **get_and()** which uses `str.contains()` and a search (e.g. df[df['area'].str.contains()]) to filter and `return` all areas with 'And' in the name  (Note:  case is important)

**Test output**:  
105 rows × 13 columns


In [ ]:

def get_and():
# add code to return just rows in which area contains 'And'
  area_cap_and = housing_df[housing_df['area'].str.title().str.contains('And')]
  return area_cap_and





# run and test if returned is correct length 

actual = len(get_and())
expected = 105

if actual == expected:
  print("Test passed", actual)
else:
  print("Test failed, expected", expected, "got", actual)

Test passed 105


### Exercise 5
---

Filter the data for all areas starting with 'Ba'  

*hint: use `startswith()`*

**Test Ouput:**  
42 rows, first row has area 'Barking and Dagenham'

In [ ]:
def get_ba():
  # add code to filter for all areas starting with 'Ba'
  start_ba = housing_df[housing_df['area'].str.startswith('ba')]
  return start_ba

get_ba()


#run and test if your returned rows are the right length 

actual = len(get_ba())
expected = 42 

if actual == expected:
  print("Test passed", actual)
else:
  print("Test failed, expected", expected, "got", actual)


Test passed 42


### Exercise 6
---
Create function called **get_ham()** to filter and `return` the data for all areas ending with 'ham', for the year 2000

*hint: use `endswith()`*   

**Test output**:  
4 rows (barking and dagenham, hammersmith and fulham, lewisham, newham)  

In [ ]:
def get_ham():
  # add code to return rows which end with 'ham' for the year 2000
  end_ham = housing_df[(housing_df['area'].str.endswith('ham')) & (housing_df['date'].str.split('-').str.get(0) == '2000')]
  return end_ham
  display(end_ham)
  

get_ham()

#run and test if correct number of rows are returned

actual = len(get_ham())
expected = 4 

if actual == expected:
  print("Test passed", actual)
else:
  print("Test failed, expected", expected, "got", actual)

Test passed 4


### Exercise 7 - new data set
---

Use the data set here:  https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true

Read the data from the sheet 'Skill Migration'  

Write a function called **create_new_df()** which will inspect the data, then create and return new dataframe with the following changes:

1.  Remove the word 'Skills' from the 'skill_group_category' column   
  *hint: you can use the `str.rstrip()` function*
2.  Convert country_code to uppercase    
  *hint: try `upper()`*  
4.  Remove the skill_group_id and the wb_income columns
3.  Filter for regions containing 'Asia'  
  *hint: you might have to `return` it*

**Test output**:  
9969 rows × 10 columns

In [15]:
import pandas as pd

def create_new_df():
  # add code below to return a df with 'Skills' removed, country_code in uppercase, no skill_group_id or wb_income columns and only for regions containing Asia 
  skill_migration = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true', sheet_name = 'Skill Migration')
  print(skill_migration.columns)
  print(skill_migration['skill_group_category'].unique())
  skill_migration['skill_group_category'] = skill_migration['skill_group_category'].str.rstrip('Skills')
  display(skill_migration)
  skill_migration['country_code'] = skill_migration['country_code'].str.upper()
  display(skill_migration)
  skill_migration = skill_migration.drop(['skill_group_id', 'wb_income'],1)
  print(skill_migration.columns)
  skill_migration = skill_migration[skill_migration['wb_region'].str.contains('Asia')]
  display(skill_migration)
  return skill_migration

create_new_df()



#run and test if returned dataframe is correct length, with the right number of columns  and first row skill_group_category is correct 
test_df = create_new_df()
actual_len = len(test_df)
actual_col = len(test_df.columns)
expected_len = 9969
expected_col = 10
actual_skill = test_df['skill_group_category'].iloc[0]
expected_skill = 'Tech '

if actual_len == expected_len and actual_col == expected_col and actual_skill == expected_skill:
  print("Test passed", actual_len, "x", actual_col, actual_skill)
else:
  print("Test failed, expected", expected_len, "x", expected_col, expected_skill, "got", actual_len, "x", actual_col, actual_skill)

Index(['country_code', 'country_name', 'wb_income', 'wb_region',
       'skill_group_id', 'skill_group_category', 'skill_group_name',
       'net_per_10K_2015', 'net_per_10K_2016', 'net_per_10K_2017',
       'net_per_10K_2018', 'net_per_10K_2019'],
      dtype='object')
['Tech Skills' 'Business Skills' 'Specialized Industry Skills'
 'Disruptive Tech Skills' 'Soft Skills']


country_code country_name   wb_income            wb_region  \
0               af  Afghanistan  Low income           South Asia   
1               af  Afghanistan  Low income           South Asia   
2               af  Afghanistan  Low income           South Asia   
3               af  Afghanistan  Low income           South Asia   
4               af  Afghanistan  Low income           South Asia   
...            ...          ...         ...                  ...   
17612           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17613           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17614           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17615           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17616           zw     Zimbabwe  Low income  Sub-Saharan Africa    

       skill_group_id   skill_group_category        skill_group_name  \
0                2549                  Tech   Information Management   
1                2608              Business   Operational Efficiency   
2                3806  Specialized Industry        National Security   
3               50321                  Tech         Software Testing   
4                1606  Specialized Industry                     Navy   
...               ...                    ...                     ...   
17612           12666  Specialized Industry                 Teaching   
17613            1235  Specialized Industry                   Mining   
17614           43756  Specialized Industry        Personal Coaching   
17615            1724  Specialized Industry            Public Health   
17616             172  Specialized Industry                      Law   

       net_per_10K_2015  net_per_10K_2016  net_per_10K_2017  net_per_10K_2018  \
0               -791.59           -705.88           -550.04           -680.92   
1              -1610.25           -933.55           -776.06           -532.22   
2              -1731.45           -769.68           -756.59           -600.44   
3               -957.50           -828.54           -964.73           -406.50   
4              -1510.71           -841.17           -842.32           -581.71   
...                 ...               ...               ...               ...   
17612             71.18             30.68            -18.85            -68.89   
17613              8.97           -112.85            -35.87            -65.38   
17614            -53.45            -59.70            -88.01            -55.90   
17615             15.25            -65.53            -57.22            -39.39   
17616             40.85             47.16            -31.24             -8.24   

       net_per_10K_2019  
0              -1208.79  
1               -790.09  
2               -767.64  
3               -739.51  
4               -718.64  
...                 ...  
17612            -93.70  
17613            -93.46  
17614            -82.23  
17615            -32.14  
17616            -25.52  

[17617 rows x 12 columns]

country_code country_name   wb_income            wb_region  \
0               AF  Afghanistan  Low income           South Asia   
1               AF  Afghanistan  Low income           South Asia   
2               AF  Afghanistan  Low income           South Asia   
3               AF  Afghanistan  Low income           South Asia   
4               AF  Afghanistan  Low income           South Asia   
...            ...          ...         ...                  ...   
17612           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17613           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17614           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17615           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17616           ZW     Zimbabwe  Low income  Sub-Saharan Africa    

       skill_group_id   skill_group_category        skill_group_name  \
0                2549                  Tech   Information Management   
1                2608              Business   Operational Efficiency   
2                3806  Specialized Industry        National Security   
3               50321                  Tech         Software Testing   
4                1606  Specialized Industry                     Navy   
...               ...                    ...                     ...   
17612           12666  Specialized Industry                 Teaching   
17613            1235  Specialized Industry                   Mining   
17614           43756  Specialized Industry        Personal Coaching   
17615            1724  Specialized Industry            Public Health   
17616             172  Specialized Industry                      Law   

       net_per_10K_2015  net_per_10K_2016  net_per_10K_2017  net_per_10K_2018  \
0               -791.59           -705.88           -550.04           -680.92   
1              -1610.25           -933.55           -776.06           -532.22   
2              -1731.45           -769.68           -756.59           -600.44   
3               -957.50           -828.54           -964.73           -406.50   
4              -1510.71           -841.17           -842.32           -581.71   
...                 ...               ...               ...               ...   
17612             71.18             30.68            -18.85            -68.89   
17613              8.97           -112.85            -35.87            -65.38   
17614            -53.45            -59.70            -88.01            -55.90   
17615             15.25            -65.53            -57.22            -39.39   
17616             40.85             47.16            -31.24             -8.24   

       net_per_10K_2019  
0              -1208.79  
1               -790.09  
2               -767.64  
3               -739.51  
4               -718.64  
...                 ...  
17612            -93.70  
17613            -93.46  
17614            -82.23  
17615            -32.14  
17616            -25.52  

[17617 rows x 12 columns]

Index(['country_code', 'country_name', 'wb_region', 'skill_group_category',
       'skill_group_name', 'net_per_10K_2015', 'net_per_10K_2016',
       'net_per_10K_2017', 'net_per_10K_2018', 'net_per_10K_2019'],
      dtype='object')


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  if sys.path[0] == '':


country_code country_name            wb_region   skill_group_category  \
0               AF  Afghanistan           South Asia                  Tech    
1               AF  Afghanistan           South Asia              Business    
2               AF  Afghanistan           South Asia  Specialized Industry    
3               AF  Afghanistan           South Asia                  Tech    
4               AF  Afghanistan           South Asia  Specialized Industry    
...            ...          ...                  ...                    ...   
17371           VN      Vietnam  East Asia & Pacific  Specialized Industry    
17372           VN      Vietnam  East Asia & Pacific  Specialized Industry    
17373           VN      Vietnam  East Asia & Pacific              Business    
17374           VN      Vietnam  East Asia & Pacific       Disruptive Tech    
17375           VN      Vietnam  East Asia & Pacific              Business    

             skill_group_name  net_per_10K_2015  net_per_10K_2016  \
0      Information Management           -791.59           -705.88   
1      Operational Efficiency          -1610.25           -933.55   
2           National Security          -1731.45           -769.68   
3            Software Testing           -957.50           -828.54   
4                        Navy          -1510.71           -841.17   
...                       ...               ...               ...   
17371               Air Force            658.89            291.89   
17372                 Cooking             83.81            236.02   
17373             Sales Leads            511.16             98.13   
17374   Aerospace Engineering            410.02             42.71   
17375        Revenue Analysis            315.68            158.12   

       net_per_10K_2017  net_per_10K_2018  net_per_10K_2019  
0               -550.04           -680.92          -1208.79  
1               -776.06           -532.22           -790.09  
2               -756.59           -600.44           -767.64  
3               -964.73           -406.50           -739.51  
4               -842.32           -581.71           -718.64  
...                 ...               ...               ...  
17371           -271.00            219.20            166.99  
17372            -42.85            238.90            180.90  
17373           -355.91            133.89            181.47  
17374           -222.09            138.03            256.55  
17375            101.89            277.98            267.58  

[9969 rows x 10 columns]

Index(['country_code', 'country_name', 'wb_income', 'wb_region',
       'skill_group_id', 'skill_group_category', 'skill_group_name',
       'net_per_10K_2015', 'net_per_10K_2016', 'net_per_10K_2017',
       'net_per_10K_2018', 'net_per_10K_2019'],
      dtype='object')
['Tech Skills' 'Business Skills' 'Specialized Industry Skills'
 'Disruptive Tech Skills' 'Soft Skills']


country_code country_name   wb_income            wb_region  \
0               af  Afghanistan  Low income           South Asia   
1               af  Afghanistan  Low income           South Asia   
2               af  Afghanistan  Low income           South Asia   
3               af  Afghanistan  Low income           South Asia   
4               af  Afghanistan  Low income           South Asia   
...            ...          ...         ...                  ...   
17612           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17613           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17614           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17615           zw     Zimbabwe  Low income  Sub-Saharan Africa    
17616           zw     Zimbabwe  Low income  Sub-Saharan Africa    

       skill_group_id   skill_group_category        skill_group_name  \
0                2549                  Tech   Information Management   
1                2608              Business   Operational Efficiency   
2                3806  Specialized Industry        National Security   
3               50321                  Tech         Software Testing   
4                1606  Specialized Industry                     Navy   
...               ...                    ...                     ...   
17612           12666  Specialized Industry                 Teaching   
17613            1235  Specialized Industry                   Mining   
17614           43756  Specialized Industry        Personal Coaching   
17615            1724  Specialized Industry            Public Health   
17616             172  Specialized Industry                      Law   

       net_per_10K_2015  net_per_10K_2016  net_per_10K_2017  net_per_10K_2018  \
0               -791.59           -705.88           -550.04           -680.92   
1              -1610.25           -933.55           -776.06           -532.22   
2              -1731.45           -769.68           -756.59           -600.44   
3               -957.50           -828.54           -964.73           -406.50   
4              -1510.71           -841.17           -842.32           -581.71   
...                 ...               ...               ...               ...   
17612             71.18             30.68            -18.85            -68.89   
17613              8.97           -112.85            -35.87            -65.38   
17614            -53.45            -59.70            -88.01            -55.90   
17615             15.25            -65.53            -57.22            -39.39   
17616             40.85             47.16            -31.24             -8.24   

       net_per_10K_2019  
0              -1208.79  
1               -790.09  
2               -767.64  
3               -739.51  
4               -718.64  
...                 ...  
17612            -93.70  
17613            -93.46  
17614            -82.23  
17615            -32.14  
17616            -25.52  

[17617 rows x 12 columns]

country_code country_name   wb_income            wb_region  \
0               AF  Afghanistan  Low income           South Asia   
1               AF  Afghanistan  Low income           South Asia   
2               AF  Afghanistan  Low income           South Asia   
3               AF  Afghanistan  Low income           South Asia   
4               AF  Afghanistan  Low income           South Asia   
...            ...          ...         ...                  ...   
17612           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17613           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17614           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17615           ZW     Zimbabwe  Low income  Sub-Saharan Africa    
17616           ZW     Zimbabwe  Low income  Sub-Saharan Africa    

       skill_group_id   skill_group_category        skill_group_name  \
0                2549                  Tech   Information Management   
1                2608              Business   Operational Efficiency   
2                3806  Specialized Industry        National Security   
3               50321                  Tech         Software Testing   
4                1606  Specialized Industry                     Navy   
...               ...                    ...                     ...   
17612           12666  Specialized Industry                 Teaching   
17613            1235  Specialized Industry                   Mining   
17614           43756  Specialized Industry        Personal Coaching   
17615            1724  Specialized Industry            Public Health   
17616             172  Specialized Industry                      Law   

       net_per_10K_2015  net_per_10K_2016  net_per_10K_2017  net_per_10K_2018  \
0               -791.59           -705.88           -550.04           -680.92   
1              -1610.25           -933.55           -776.06           -532.22   
2              -1731.45           -769.68           -756.59           -600.44   
3               -957.50           -828.54           -964.73           -406.50   
4              -1510.71           -841.17           -842.32           -581.71   
...                 ...               ...               ...               ...   
17612             71.18             30.68            -18.85            -68.89   
17613              8.97           -112.85            -35.87            -65.38   
17614            -53.45            -59.70            -88.01            -55.90   
17615             15.25            -65.53            -57.22            -39.39   
17616             40.85             47.16            -31.24             -8.24   

       net_per_10K_2019  
0              -1208.79  
1               -790.09  
2               -767.64  
3               -739.51  
4               -718.64  
...                 ...  
17612            -93.70  
17613            -93.46  
17614            -82.23  
17615            -32.14  
17616            -25.52  

[17617 rows x 12 columns]

Index(['country_code', 'country_name', 'wb_region', 'skill_group_category',
       'skill_group_name', 'net_per_10K_2015', 'net_per_10K_2016',
       'net_per_10K_2017', 'net_per_10K_2018', 'net_per_10K_2019'],
      dtype='object')


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:12: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  if sys.path[0] == '':


country_code country_name            wb_region   skill_group_category  \
0               AF  Afghanistan           South Asia                  Tech    
1               AF  Afghanistan           South Asia              Business    
2               AF  Afghanistan           South Asia  Specialized Industry    
3               AF  Afghanistan           South Asia                  Tech    
4               AF  Afghanistan           South Asia  Specialized Industry    
...            ...          ...                  ...                    ...   
17371           VN      Vietnam  East Asia & Pacific  Specialized Industry    
17372           VN      Vietnam  East Asia & Pacific  Specialized Industry    
17373           VN      Vietnam  East Asia & Pacific              Business    
17374           VN      Vietnam  East Asia & Pacific       Disruptive Tech    
17375           VN      Vietnam  East Asia & Pacific              Business    

             skill_group_name  net_per_10K_2015  net_per_10K_2016  \
0      Information Management           -791.59           -705.88   
1      Operational Efficiency          -1610.25           -933.55   
2           National Security          -1731.45           -769.68   
3            Software Testing           -957.50           -828.54   
4                        Navy          -1510.71           -841.17   
...                       ...               ...               ...   
17371               Air Force            658.89            291.89   
17372                 Cooking             83.81            236.02   
17373             Sales Leads            511.16             98.13   
17374   Aerospace Engineering            410.02             42.71   
17375        Revenue Analysis            315.68            158.12   

       net_per_10K_2017  net_per_10K_2018  net_per_10K_2019  
0               -550.04           -680.92          -1208.79  
1               -776.06           -532.22           -790.09  
2               -756.59           -600.44           -767.64  
3               -964.73           -406.50           -739.51  
4               -842.32           -581.71           -718.64  
...                 ...               ...               ...  
17371           -271.00            219.20            166.99  
17372            -42.85            238.90            180.90  
17373           -355.91            133.89            181.47  
17374           -222.09            138.03            256.55  
17375            101.89            277.98            267.58  

[9969 rows x 10 columns]

Test passed 9969 x 10 Tech 


### Exercise 8
---

Write a function called **clean_skills()** that will:
1. rename the **net_per_10K_year** columns to be just the year
2. in the **skill_group_category** column replace the 'z' in 'specialized' with 's' to Anglicise the spelling. 

The function should `return` the cleaned data.  

Hint:  You can use the `replace()` function to replace substring's and characters in both column headings and the actual data.  
* `.str.replace("old","new")`

**Test output**:  
17617 rows × 12 columns, with z replace by s in Specialized  
Column names: country_code	country_name	wb_income	wb_region	skill_group_id	skill_group_category	skill_group_name	2015	2016	2017	2018	2019

In [54]:
import pandas as pd
migration = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true', sheet_name = 'Skill Migration')# add code to open dataset here

#create a list of the column names in migration that start with 'net_per_10K'
#for each column header in tern, split it by '_' and take the year
#in migration df, if the column header ends with that year, replace the whole header with the year

def clean_skills(df):
  net_cols = [col for col in df.columns if col.startswith('net_per_10K')]
  for col in net_cols:
    col_year = col.split('_')[3]
    if col.endswith(col_year):
      df = df.rename(columns = {col:col_year})
  print(df.columns)
  migration['skill_group_category'] = migration['skill_group_category'].str.replace('Specialized', 'Specialised')
  return df

clean_skills(migration)

# run and test if columns have correct names and specialised is anglised 
test_df = clean_skills(migration)

if (test_df['skill_group_category'].str.contains('Specialised').any() == True) and (test_df.columns.str.contains('net_per_10K_').any() == False):
  print("Test passed")
else:
  print("Test failed")


Index(['country_code', 'country_name', 'wb_income', 'wb_region',
       'skill_group_id', 'skill_group_category', 'skill_group_name', '2015',
       '2016', '2017', '2018', '2019'],
      dtype='object')
Index(['country_code', 'country_name', 'wb_income', 'wb_region',
       'skill_group_id', 'skill_group_category', 'skill_group_name', '2015',
       '2016', '2017', '2018', '2019'],
      dtype='object')
Test passed


### Exercise 9
---

Read the 'Country Migration' sheet.

Write a function that will:  
*  convert the country codes to upper case  
*  drop the lat and long columns for both base and target  
*  rename the net_per_10K_year columns to year only  
*  filter for base_country_wb_region contains 'Africa' and target_country_wb_region contains Asia  

**Test output**:  
```
base_country_code	base_country_name	base_country_wb_income	base_country_wb_region	target_country_code	target_country_name	target_country_wb_income	target_country_wb_region	2015	2016	2017	2018	2019
0	AE	United Arab Emirates	High Income	Middle East & North Africa	AF	Afghanistan	Low Income	South Asia	0.19	0.16	0.11	-0.05	-0.02
4	AE	United Arab Emirates	High Income	Middle East & North Africa	AM	Armenia	Upper Middle Income	Europe & Central Asia	0.10	0.05	0.03	-0.01	0.02
5	AE	United Arab Emirates	High Income	Middle East & North Africa	AU	Australia	High Income	East Asia & Pacific	-1.06	-3.31	-4.01	-4.58	-4.09
6	AE	United Arab Emirates	High Income	Middle East & North Africa	AT	Austria	High Income	Europe & Central Asia	0.11	-0.08	-0.07	-0.05	-0.16
7	AE	United Arab Emirates	High Income	Middle East & North Africa	AZ	Azerbaijan	Upper Middle Income	Europe & Central Asia	0.24	0.25	0.10	0.05	0.04
...	...	...	...	...	...	...	...	...	...	...	...	...	...
4132	ZM	Zambia	Lower Middle Income	Sub-Saharan Africa	GB	United Kingdom	High Income	Europe & Central Asia	43.27	27.60	7.88	6.90	3.68
4135	ZW	Zimbabwe	Low Income	Sub-Saharan Africa	AU	Australia	High Income	East Asia & Pacific	-1.31	-2.33	-2.10	-2.08	-1.84
4138	ZW	Zimbabwe	Low Income	Sub-Saharan Africa	IS	Iceland	High Income	Europe & Central Asia	8.52	6.22	2.35	1.81	0.97
4142	ZW	Zimbabwe	Low Income	Sub-Saharan Africa	NO	Norway	High Income	Europe & Central Asia	2.88	6.46	2.10	0.33	-0.13
4145	ZW	Zimbabwe	Low Income	Sub-Saharan Africa	GB	United Kingdom	High Income	Europe & Central Asia	3.91	4.66	0.74	-0.66	-1.97
478 rows × 13 columns
```



In [71]:
def clean_country_mig():
  # add code below to clean the data 
  country_migration = pd.read_excel('https://github.com/futureCodersSE/working-with-data/blob/main/Data%20sets/public_use-talent-migration.xlsx?raw=true', sheet_name = 'Country Migration')
  #country_migration.info()
  country_migration['base_country_code'] = country_migration['base_country_code'].str.upper()
  country_migration['target_country_code'] = country_migration['target_country_code'].str.upper()
  #print(country_migration.head())
  country_migration = country_migration.drop(['base_lat', 'base_long', 'target_lat', 'target_long'],1)
  #print(country_migration.columns)
  net_cols = [col for col in country_migration.columns if col.startswith('net_per_10K')]
  for col in net_cols:
    col_year = col.split('_')[3]
    if col.endswith(col_year):
      country_migration = country_migration.rename(columns = {col:col_year})
  #print(country_migration.columns)
  country_migration = country_migration[country_migration['base_country_wb_region'].str.contains('Africa') & country_migration['target_country_wb_region'].str.contains('Asia')]
  #print(country_migration.head())
  return country_migration

clean_country_mig()


# run test if there is the correct number of columns, country codes are in uppercase and year columns have been reformatted 

test_df = clean_country_mig()
actual_col_len = len(test_df.columns)
expected = 13

if actual_col_len == expected and (test_df['base_country_code'].str.islower().any() == False) and (test_df.columns.str.contains('net_per_10K_').any() == False):
  print("Test passed")
else:
  print("Test failed")

/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  


Test passed


### Exercise 10
---

Read the data from file 'https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv'.

Write a function that will return a new dataframe with just the married women listed, surname only.

**Test output**:  
```
	PassengerId	Survived	Pclass	Name	Sex	Age	SibSp	Parch	Ticket	Fare	Cabin	Embarked
1	2	1	1	Cumings	female	38.0	1	0	PC 17599	71.2833	C85	C
3	4	1	1	Futrelle	female	35.0	1	0	113803	53.1000	C123	S
8	9	1	3	Johnson	female	27.0	0	2	347742	11.1333	NaN	S
9	10	1	2	Nasser	female	14.0	1	0	237736	30.0708	NaN	C
15	16	1	2	Hewlett	female	55.0	0	0	248706	16.0000	NaN	S
...	...	...	...	...	...	...	...	...	...	...	...	...
871	872	1	1	Beckwith	female	47.0	1	1	11751	52.5542	D35	S
874	875	1	2	Abelson	female	28.0	1	0	P/PP 3381	24.0000	NaN	C
879	880	1	1	Potter	female	56.0	0	1	11767	83.1583	C50	C
880	881	1	2	Shelley	female	25.0	0	1	230433	26.0000	NaN	S
885	886	0	3	Rice	female	39.0	0	5	382652	29.1250	NaN	Q
129 rows × 12 columns
```





In [89]:
titanic = pd.read_csv('https://raw.githubusercontent.com/pandas-dev/pandas/master/doc/data/titanic.csv')

#print(titanic.head())

def get_married(df):
  #add code to return only the last names of married women
  df = df[df['Name'].str.contains('Mrs')]
  df['Name'] = df['Name'].str.split(',').str.get(0)
  df = df[df['Sex'] == 'female']
  #print(df.head())
  return df

get_married(titanic)

# run and test if returned dataframe is correct length and has correct first row 
test_df = get_married(titanic)
actual_len = len(test_df)
expected_len = 129
actual_name = test_df['Name'].iloc[0]
expected_name = 'Cumings'

if actual_len == expected_len and actual_name == expected_name:
  print("Test passed, ", actual_len, actual_name)
else:
  print("Test failed expected ", expected_len, expected_name, "got", actual_len, actual_name)

Test passed,  129 Cumings


/usr/local/lib/python3.7/dist-packages/ipykernel_launcher.py:8: SettingWithCopyWarning: 
A value is trying to be set on a copy of a slice from a DataFrame.
Try using .loc[row_indexer,col_indexer] = value instead

See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy
  


---
### Optional extra practice

There are some similar and some more challenging exercises [here](https://www.w3resource.com/python-exercises/date-time-exercise/) if you would like to practice more. The site has its own editor.

# Reflection
----

## What skills have you demonstrated in completing this notebook?

Your answer:

## What caused you the most difficulty?

Your answer: